File: 0213-house-robber-ii/0213-house-robber-ii.cpp
class Solution {
public:
    int f(vector<int> nums,int ind){
        // if(ind==0) return nums[0];

        // if(dp[ind]!=-1) return dp[ind];
        // int pick=nums[ind];
        // if(ind>1) pick=nums[ind]+f(nums,ind-2,dp);
        // int notpick=0+f(nums,ind-1,dp);
        // return dp[ind]=max(pick,notpick);

        int prev=nums[0],prev2=0;
        for(int i=0;i<nums.size();i++){
            int pick = nums[i];
            if(i>1) pick=pick+prev2;
            int notpick=prev+0;
            int curi=max(pick,notpick);
            prev2=prev;
            prev=curi;
        }
        return prev;
    }
    int rob(vector<int>& nums) {vector<int> x,y;
        // vector<int> dp(nums.size()-1,-1);
        if(nums.size()==1) return nums[0];
        for(int i=0;i<nums.size();i++){
            if(i!=0) x.push_back(nums[i]);
            if(i!=nums.size()-1) y.push_back(nums[i]);
        }
        return max(f(x,x.size()-1),f(y,y.size()-1));
    }
};

File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        if(nums.length==1) return nums[0];
        ArrayList<Integer> nums1 =  new ArrayList<Integer>();
        ArrayList<Integer> nums2 =  new ArrayList<Integer>();
        for(int i = 0;i<nums.length;i++){
            if(i!=0){
                nums1.add(nums[i]);
            }
            if(i!=nums.length-1){
                nums2.add(nums[i]);
            }
        }
        return Math.max(fun(nums1,nums1.size()-1),fun(nums2,nums2.size()-1));
    }
    public int fun(ArrayList<Integer> nums,int ind){
        int prev = nums.get(0);
        int prev2 = 0;
        
        for(int i = 1;i<nums.size();i++){
            int take = nums.get(i);
            if(i>1) take = nums.get(i) + prev2;
            int notake = prev;
            int cur = Math.max(take,notake);
            prev2 = prev;
            prev  = cur;
        }
        return prev;
    }
}